charts/redis-ha/values.yaml

```diff
@@ -306,6 +310,13 @@ securityContext:
   fsGroup: 1000
   runAsNonRoot: true
 
+  ## Assuming your kubelet allows it, you can the following instructions to configure
+  ## specific sysctl parameters
+  ##
+  # sysctls:
+  # - name: net.core.somaxconn
+  #   value: '10000'
```

Comment:
Does helm still set these parameters if you use this as a dependency to another chart? 

I don't think it does...

Reply:
👋 I double checked this and I can confirm that it works as expected:

<details>
  <summary>test subchart</summary>
  
  ```diff
~$ git diff
diff --git a/charts/test/Chart.yaml b/charts/test/Chart.yaml
new file mode 100644
index 0000000..5d20db0
--- /dev/null
+++ b/charts/test/Chart.yaml
@@ -0,0 +1,5 @@
+---
+apiVersion: v1
+name: test
+version: 0.0.0
+appVersion: 0.0.0
diff --git a/charts/test/charts/redis-ha-4.12.2.tgz b/charts/test/charts/redis-ha-4.12.2.tgz
new file mode 100644
index 0000000..88e40dd
Binary files /dev/null and b/charts/test/charts/redis-ha-4.12.2.tgz differ
diff --git a/charts/test/requirements.lock b/charts/test/requirements.lock
new file mode 100644
index 0000000..2fdfcc1
--- /dev/null
+++ b/charts/test/requirements.lock
@@ -0,0 +1,6 @@
+dependencies:
+- name: redis-ha
+  repository: file://../redis-ha
+  version: 4.12.2
+digest: sha256:724e560280badb99c2889472d25edb68acb3ee4db92119fa6e9666b096eeda7e
+generated: "2021-02-18T09:39:21.173729Z"
diff --git a/charts/test/requirements.yaml b/charts/test/requirements.yaml
new file mode 100644
index 0000000..3a2d70d
--- /dev/null
+++ b/charts/test/requirements.yaml
@@ -0,0 +1,4 @@
+dependencies:
+  - name: redis-ha
+    version: 0.0.0
+    repository: file://../redis-ha
diff --git a/charts/test/values.yaml b/charts/test/values.yaml
new file mode 100644
index 0000000..c16d472
--- /dev/null
+++ b/charts/test/values.yaml
@@ -0,0 +1,9 @@
+---
+redis-ha:
+  podSecurityPolicy:
+    create: true
+
+  securityContext:
+    sysctls:
+      - name: net.core.somaxconn
+        value: '10000'
\ No newline at end of file
  ```
</details>

<details>
  <summary>results</summary>
  
  ```diff
~$ helm diff upgrade --allow-unreleased --debug test ./charts/test
backend, test-redis-ha-announce-2, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-announce-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha-announce-2
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
+ spec:
+   publishNotReadyAddresses: true
+   type: ClusterIP
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
+     "statefulset.kubernetes.io/pod-name": test-redis-ha-server-2
backend, test-redis-ha, ServiceAccount (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-serviceaccount.yaml
+ apiVersion: v1
+ kind: ServiceAccount
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     heritage: Helm
+     release: test
+     chart: redis-ha-4.12.2
+     app: test-redis-ha
backend, test-redis-ha-configmap, ConfigMap (v1) has been added:
+ # Source: test/charts/redis-ha/templates/redis-ha-configmap.yaml
+ apiVersion: v1
+ kind: ConfigMap
+ metadata:
+   name: test-redis-ha-configmap
+   namespace: backend
+   labels:
+     heritage: Helm
+     release: test
+     chart: redis-ha-4.12.2
+     app: test-redis-ha
+ data:
+   redis.conf: |
+     dir "/data"
+     port 6379
+     maxmemory 0
+     maxmemory-policy volatile-lru
+     min-replicas-max-lag 5
+     min-replicas-to-write 1
+     rdbchecksum yes
+     rdbcompression yes
+     repl-diskless-sync yes
+     save 900 1
+
+   sentinel.conf: |
+     dir "/data"
+     port 26379
+         sentinel down-after-milliseconds mymaster 10000
+         sentinel failover-timeout mymaster 180000
+         maxclients 10000
+         sentinel parallel-syncs mymaster 5
+
+   init.sh: |
+     echo "$(date) Start..."
+     HOSTNAME="$(hostname)"
+     INDEX="${HOSTNAME##*-}"
+     SENTINEL_PORT=26379
+     MASTER=''
+     MASTER_GROUP="mymaster"
+     QUORUM="2"
+     REDIS_CONF=/data/conf/redis.conf
+     REDIS_PORT=6379
+     REDIS_TLS_PORT=
+     SENTINEL_CONF=/data/conf/sentinel.conf
+     SENTINEL_TLS_PORT=
+     SERVICE=test-redis-ha
+     SENTINEL_TLS_REPLICATION_ENABLED=false
+     REDIS_TLS_REPLICATION_ENABLED=false
+     set -eu
+
+     sentinel_get_master() {
+     set +e
+         if [ "$SENTINEL_PORT" -eq 0 ]; then
+             redis-cli -h "${SERVICE}" -p "${SENTINEL_TLS_PORT}"   --tls --cacert /tls-certs/ca.crt --cert /tls-certs/redis.crt --key /tls-certs/redis.key sentinel get-master-addr-by-name "${MASTER_GROUP}" |\
+             grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}'
+         else
+             redis-cli -h "${SERVICE}" -p "${SENTINEL_PORT}"  sentinel get-master-addr-by-name "${MASTER_GROUP}" |\
+             grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}'
+         fi
+     set -e
+     }
+
+     sentinel_get_master_retry() {
+         master=''
+         retry=${1}
+         sleep=3
+         for i in $(seq 1 "${retry}"); do
+             master=$(sentinel_get_master)
+             if [ -n "${master}" ]; then
+                 break
+             fi
+             sleep $((sleep + i))
+         done
+         echo "${master}"
+     }
+
+     identify_master() {
+         echo "Identifying redis master (get-master-addr-by-name).."
+         echo "  using sentinel (test-redis-ha), sentinel group name (mymaster)"
+         echo "  $(date).."
+         MASTER="$(sentinel_get_master_retry 3)"
+         if [ -n "${MASTER}" ]; then
+             echo "  $(date) Found redis master (${MASTER})"
+         else
+             echo "  $(date) Did not find redis master (${MASTER})"
+         fi
+     }
+
+     sentinel_update() {
+         echo "Updating sentinel config.."
+         echo "  evaluating sentinel id (\${SENTINEL_ID_${INDEX}})"
+         eval MY_SENTINEL_ID="\$SENTINEL_ID_${INDEX}"
+         echo "  sentinel id (${MY_SENTINEL_ID}), sentinel grp (${MASTER_GROUP}), quorum (${QUORUM})"
+         sed -i "1s/^/sentinel myid ${MY_SENTINEL_ID}\\n/" "${SENTINEL_CONF}"
+         if [ "$SENTINEL_TLS_REPLICATION_ENABLED" = true ]; then
+             echo "  redis master (${1}:${REDIS_TLS_PORT})"
+             sed -i "2s/^/sentinel monitor ${MASTER_GROUP} ${1} ${REDIS_TLS_PORT} ${QUORUM} \\n/" "${SENTINEL_CONF}"
+         else
+             echo "  redis master (${1}:${REDIS_PORT})"
+             sed -i "2s/^/sentinel monitor ${MASTER_GROUP} ${1} ${REDIS_PORT} ${QUORUM} \\n/" "${SENTINEL_CONF}"
+         fi
+         echo "sentinel announce-ip ${ANNOUNCE_IP}" >> ${SENTINEL_CONF}
+         if [ "$SENTINEL_PORT" -eq 0 ]; then
+             echo "  announce (${ANNOUNCE_IP}:${SENTINEL_TLS_PORT})"
+             echo "sentinel announce-port ${SENTINEL_TLS_PORT}" >> ${SENTINEL_CONF}
+         else
+             echo "  announce (${ANNOUNCE_IP}:${SENTINEL_PORT})"
+             echo "sentinel announce-port ${SENTINEL_PORT}" >> ${SENTINEL_CONF}
+         fi
+     }
+
+     redis_update() {
+         echo "Updating redis config.."
+         if [ "$REDIS_TLS_REPLICATION_ENABLED" = true ]; then
+             echo "  we are slave of redis master (${1}:${REDIS_TLS_PORT})"
+             echo "slaveof ${1} ${REDIS_TLS_PORT}" >> "${REDIS_CONF}"
+             echo "slave-announce-port ${REDIS_TLS_PORT}" >> ${REDIS_CONF}
+         else
+             echo "  we are slave of redis master (${1}:${REDIS_PORT})"
+             echo "slaveof ${1} ${REDIS_PORT}" >> "${REDIS_CONF}"
+             echo "slave-announce-port ${REDIS_PORT}" >> ${REDIS_CONF}
+         fi
+         echo "slave-announce-ip ${ANNOUNCE_IP}" >> ${REDIS_CONF}
+     }
+
+     copy_config() {
+         echo "Copying default redis config.."
+         echo "  to '${REDIS_CONF}'"
+         cp /readonly-config/redis.conf "${REDIS_CONF}"
+         echo "Copying default sentinel config.."
+         echo "  to '${SENTINEL_CONF}'"
+         cp /readonly-config/sentinel.conf "${SENTINEL_CONF}"
+     }
+
+     setup_defaults() {
+         echo "Setting up defaults.."
+         echo "  using statefulset index (${INDEX})"
+         if [ "${INDEX}" = "0" ]; then
+             echo "Setting this pod as master for redis and sentinel.."
+             echo "  using announce (${ANNOUNCE_IP})"
+             redis_update "${ANNOUNCE_IP}"
+             sentinel_update "${ANNOUNCE_IP}"
+             echo "  make sure ${ANNOUNCE_IP} is not a slave (slaveof no one)"
+             sed -i "s/^.*slaveof.*//" "${REDIS_CONF}"
+         else
+             echo "Getting redis master ip.."
+             echo "  blindly assuming (${SERVICE}-announce-0) or (${SERVICE}-server-0) are master"
+             DEFAULT_MASTER="$(getent_hosts 0 | awk '{ print $1 }')"
+             echo "  identified redis (may be redis master) ip (${DEFAULT_MASTER})"
+             if [ -z "${DEFAULT_MASTER}" ]; then
+                 echo "Error: Unable to resolve redis master (getent hosts)."
+                 exit 1
+             fi
+             echo "Setting default slave config for redis and sentinel.."
+             echo "  using master ip (${DEFAULT_MASTER})"
+             redis_update "${DEFAULT_MASTER}"
+             sentinel_update "${DEFAULT_MASTER}"
+         fi
+     }
+
+     redis_ping() {
+     set +e
+         if [ "$REDIS_PORT" -eq 0 ]; then
+             redis-cli -h "${MASTER}" -p "${REDIS_TLS_PORT}"  --tls --cacert /tls-certs/ca.crt --cert /tls-certs/redis.crt --key /tls-certs/redis.key ping
+         else
+             redis-cli -h "${MASTER}" -p "${REDIS_PORT}" ping
+         fi
+     set -e
+     }
+
+     redis_ping_retry() {
+         ping=''
+         retry=${1}
+         sleep=3
+         for i in $(seq 1 "${retry}"); do
+             if [ "$(redis_ping)" = "PONG" ]; then
+                ping='PONG'
+                break
+             fi
+             sleep $((sleep + i))
+             MASTER=$(sentinel_get_master)
+         done
+         echo "${ping}"
+     }
+
+     find_master() {
+         echo "Verifying redis master.."
+         if [ "$REDIS_PORT" -eq 0 ]; then
+             echo "  ping (${MASTER}:${REDIS_TLS_PORT})"
+         else
+             echo "  ping (${MASTER}:${REDIS_PORT})"
+         fi
+         echo "  $(date).."
+         if [ "$(redis_ping_retry 3)" != "PONG" ]; then
+             echo "  $(date) Can't ping redis master (${MASTER})"
+             echo "Attempting to force failover (sentinel failover).."
+
+             if [ "$SENTINEL_PORT" -eq 0 ]; then
+                 echo "  on sentinel (${SERVICE}:${SENTINEL_TLS_PORT}), sentinel grp (${MASTER_GROUP})"
+                 echo "  $(date).."
+                 if redis-cli -h "${SERVICE}" -p "${SENTINEL_TLS_PORT}"   --tls --cacert /tls-certs/ca.crt --cert /tls-certs/redis.crt --key /tls-certs/redis.key sentinel failover "${MASTER_GROUP}" | grep -q 'NOGOODSLAVE' ; then
+                     echo "  $(date) Failover returned with 'NOGOODSLAVE'"
+                     echo "Setting defaults for this pod.."
+                     setup_defaults
+                     return 0
+                 fi
+             else
+                 echo "  on sentinel (${SERVICE}:${SENTINEL_PORT}), sentinel grp (${MASTER_GROUP})"
+                 echo "  $(date).."
+                 if redis-cli -h "${SERVICE}" -p "${SENTINEL_PORT}"  sentinel failover "${MASTER_GROUP}" | grep -q 'NOGOODSLAVE' ; then
+                     echo "  $(date) Failover returned with 'NOGOODSLAVE'"
+                     echo "Setting defaults for this pod.."
+                     setup_defaults
+                     return 0
+                 fi
+             fi
+
+             echo "Hold on for 10sec"
+             sleep 10
+             echo "We should get redis master's ip now. Asking (get-master-addr-by-name).."
+             if [ "$SENTINEL_PORT" -eq 0 ]; then
+                 echo "  sentinel (${SERVICE}:${SENTINEL_TLS_PORT}), sentinel grp (${MASTER_GROUP})"
+             else
+                 echo "  sentinel (${SERVICE}:${SENTINEL_PORT}), sentinel grp (${MASTER_GROUP})"
+             fi
+             echo "  $(date).."
+             MASTER="$(sentinel_get_master)"
+             if [ "${MASTER}" ]; then
+                 echo "  $(date) Found redis master (${MASTER})"
+                 echo "Updating redis and sentinel config.."
+                 sentinel_update "${MASTER}"
+                 redis_update "${MASTER}"
+             else
+                 echo "$(date) Error: Could not failover, exiting..."
+                 exit 1
+             fi
+         else
+             echo "  $(date) Found reachable redis master (${MASTER})"
+             echo "Updating redis and sentinel config.."
+             sentinel_update "${MASTER}"
+             redis_update "${MASTER}"
+         fi
+     }
+
+     redis_ro_update() {
+         echo "Updating read-only redis config.."
+         echo "  redis.conf set 'replica-priority 0'"
+         echo "replica-priority 0" >> ${REDIS_CONF}
+     }
+
+     getent_hosts() {
+         index=${1:-${INDEX}}
+         service="${SERVICE}-announce-${index}"
+         pod="${SERVICE}-server-${index}"
+         host=$(getent hosts "${service}")
+         if [ -z "${host}" ]; then
+             host=$(getent hosts "${pod}")
+         fi
+         echo "${host}"
+     }
+
+     mkdir -p /data/conf/
+
+     echo "Initializing config.."
+     copy_config
+
+     # where is redis master
+     identify_master
+
+     echo "Identify announce ip for this pod.."
+     echo "  using (${SERVICE}-announce-${INDEX}) or (${SERVICE}-server-${INDEX})"
+     ANNOUNCE_IP=$(getent_hosts | awk '{ print $1 }')
+     echo "  identified announce (${ANNOUNCE_IP})"
+     if [ -z "${ANNOUNCE_IP}" ]; then
+         "Error: Could not resolve the announce ip for this pod."
+         exit 1
+     elif [ "${MASTER}" ]; then
+         find_master
+     else
+         setup_defaults
+     fi
+
+     if [ "${AUTH:-}" ]; then
+         echo "Setting redis auth values.."
+         ESCAPED_AUTH=$(echo "${AUTH}" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/replace-default-auth/${ESCAPED_AUTH}/" "${REDIS_CONF}" "${SENTINEL_CONF}"
+     fi
+
+     if [ "${SENTINELAUTH:-}" ]; then
+         echo "Setting sentinel auth values"
+         ESCAPED_AUTH_SENTINEL=$(echo "$SENTINELAUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/replace-default-sentinel-auth/${ESCAPED_AUTH_SENTINEL}/" "$SENTINEL_CONF"
+     fi
+
+     echo "$(date) Ready..."
+
+   haproxy_init.sh: |
+     HAPROXY_CONF=/data/haproxy.cfg
+     cp /readonly/haproxy.cfg "$HAPROXY_CONF"
+     for loop in $(seq 1 10); do
+       getent hosts test-redis-ha-announce-0 && break
+       echo "Waiting for service test-redis-ha-announce-0 to be ready ($loop) ..." && sleep 1
+     done
+     ANNOUNCE_IP0=$(getent hosts "test-redis-ha-announce-0" | awk '{ print $1 }')
+     if [ -z "$ANNOUNCE_IP0" ]; then
+       echo "Could not resolve the announce ip for test-redis-ha-announce-0"
+       exit 1
+     fi
+     sed -i "s/REPLACE_ANNOUNCE0/$ANNOUNCE_IP0/" "$HAPROXY_CONF"
+
+     if [ "${AUTH:-}" ]; then
+         echo "Setting auth values"
+         ESCAPED_AUTH=$(echo "$AUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/REPLACE_AUTH_SECRET/${ESCAPED_AUTH}/" "$HAPROXY_CONF"
+     fi
+     for loop in $(seq 1 10); do
+       getent hosts test-redis-ha-announce-1 && break
+       echo "Waiting for service test-redis-ha-announce-1 to be ready ($loop) ..." && sleep 1
+     done
+     ANNOUNCE_IP1=$(getent hosts "test-redis-ha-announce-1" | awk '{ print $1 }')
+     if [ -z "$ANNOUNCE_IP1" ]; then
+       echo "Could not resolve the announce ip for test-redis-ha-announce-1"
+       exit 1
+     fi
+     sed -i "s/REPLACE_ANNOUNCE1/$ANNOUNCE_IP1/" "$HAPROXY_CONF"
+
+     if [ "${AUTH:-}" ]; then
+         echo "Setting auth values"
+         ESCAPED_AUTH=$(echo "$AUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/REPLACE_AUTH_SECRET/${ESCAPED_AUTH}/" "$HAPROXY_CONF"
+     fi
+     for loop in $(seq 1 10); do
+       getent hosts test-redis-ha-announce-2 && break
+       echo "Waiting for service test-redis-ha-announce-2 to be ready ($loop) ..." && sleep 1
+     done
+     ANNOUNCE_IP2=$(getent hosts "test-redis-ha-announce-2" | awk '{ print $1 }')
+     if [ -z "$ANNOUNCE_IP2" ]; then
+       echo "Could not resolve the announce ip for test-redis-ha-announce-2"
+       exit 1
+     fi
+     sed -i "s/REPLACE_ANNOUNCE2/$ANNOUNCE_IP2/" "$HAPROXY_CONF"

+     if [ "${AUTH:-}" ]; then
+         echo "Setting auth values"
+         ESCAPED_AUTH=$(echo "$AUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/REPLACE_AUTH_SECRET/${ESCAPED_AUTH}/" "$HAPROXY_CONF"
+     fi
backend, test-redis-ha-health-configmap, ConfigMap (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-health-configmap.yaml
+ apiVersion: v1
+ kind: ConfigMap
+ metadata:
+   name: test-redis-ha-health-configmap
+   namespace: backend
+   labels:
+     heritage: Helm
+     release: test
+     chart: redis-ha-4.12.2
+     app: test-redis-ha
+ data:
+   redis_liveness.sh: |
+     response=$(
+       redis-cli \
+         -h localhost \
+         -p 6379 \
+         ping
+     )
+     if [ "$response" != "PONG" ] && [ "${response:0:7}" != "LOADING" ] ; then
+       echo "$response"
+       exit 1
+     fi
+     echo "response=$response"
+   sentinel_liveness.sh: |
+     response=$(
+       redis-cli \
+         -h localhost \
+         -p 26379 \
+         ping
+     )
+     if [ "$response" != "PONG" ]; then
+       echo "$response"
+       exit 1
+     fi
+     echo "response=$response"
backend, test-redis-ha, RoleBinding (rbac.authorization.k8s.io) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-rolebinding.yaml
+ apiVersion: rbac.authorization.k8s.io/v1
+ kind: RoleBinding
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+ subjects:
+ - kind: ServiceAccount
+   name: test-redis-ha
+ roleRef:
+   apiGroup: rbac.authorization.k8s.io
+   kind: Role
+   name: test-redis-ha
backend, test-redis-ha-announce-0, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-announce-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha-announce-0
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
+ spec:
+   publishNotReadyAddresses: true
+   type: ClusterIP
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
+     "statefulset.kubernetes.io/pod-name": test-redis-ha-server-0
backend, test-redis-ha, Role (rbac.authorization.k8s.io) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-role.yaml
+ apiVersion: rbac.authorization.k8s.io/v1
+ kind: Role
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+ rules:
+ - apiGroups:
+     - ""
+   resources:
+     - endpoints
+   verbs:
+     - get
backend, test-redis-ha-announce-1, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-announce-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha-announce-1
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
+ spec:
+   publishNotReadyAddresses: true
+   type: ClusterIP
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
+     "statefulset.kubernetes.io/pod-name": test-redis-ha-server-1
backend, test-redis-ha, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+ spec:
+   type: ClusterIP
+   clusterIP: None
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
backend, test-redis-ha-server, StatefulSet (apps) has been added:
+ # Source: test/charts/redis-ha/templates/redis-ha-statefulset.yaml
+ apiVersion: apps/v1
+ kind: StatefulSet
+ metadata:
+   name: test-redis-ha-server
+   namespace: backend
+   labels:
+     test-redis-ha: replica
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     {}
+ spec:
+   selector:
+     matchLabels:
+       release: test
+       app: redis-ha
+   serviceName: test-redis-ha
+   replicas: 3
+   podManagementPolicy: OrderedReady
+   updateStrategy:
+     type: RollingUpdate
+   template:
+     metadata:
+       annotations:
+         checksum/init-config: ab774a0d7a7225b4e6dfe6337866183939293834dabef4ff427022fde6b1b669
+       labels:
+         release: test
+         app: redis-ha
+         test-redis-ha: replica
+     spec:
+       affinity:
+         podAntiAffinity:
+           requiredDuringSchedulingIgnoredDuringExecution:
+             - labelSelector:
+                 matchLabels:
+                   app: redis-ha
+                   release: test
+                   test-redis-ha: replica
+               topologyKey: kubernetes.io/hostname
+           preferredDuringSchedulingIgnoredDuringExecution:
+             - weight: 100
+               podAffinityTerm:
+                 labelSelector:
+                   matchLabels:
+                     app: redis-ha
+                     release: test
+                     test-redis-ha: replica
+                 topologyKey: failure-domain.beta.kubernetes.io/zone
+       securityContext:
+         fsGroup: 1000
+         runAsNonRoot: true
+         runAsUser: 1000
+         sysctls:
+         - name: net.core.somaxconn
+           value: "10000"
+       serviceAccountName: test-redis-ha
+       automountServiceAccountToken: false
+       initContainers:
+       - name: config-init
+         image: redis:6.0.7-alpine
+         imagePullPolicy: IfNotPresent
+         resources:
+           {}
+         command:
+         - sh
+         args:
+         - /readonly-config/init.sh
+         env:
+         - name: SENTINEL_ID_0
+           value: 9c038e829f6128e33b0ece23f256b3a7d3854756
+
+         - name: SENTINEL_ID_1
+           value: e8857b18f2db38508fe4cb5a24b54b26b07ec75c
+
+         - name: SENTINEL_ID_2
+           value: 2d63ff6edcd53314ed8cbc5d3bc9d6f660c977f0
+
+         volumeMounts:
+         - name: config
+           mountPath: /readonly-config
+           readOnly: true
+         - name: data
+           mountPath: /data
+

+       containers:
+       - name: redis
+         image: redis:6.0.7-alpine
+         imagePullPolicy: IfNotPresent
+         command:
+         - redis-server
+         args:
+         - /data/conf/redis.conf
+         livenessProbe:
+           initialDelaySeconds: 30
+           periodSeconds: 15
+           timeoutSeconds: 15
+           successThreshold: 1
+           failureThreshold: 5
+           exec:
+             command:
+               - sh
+               - -c
+               - /health/redis_liveness.sh
+         resources:
+           {}
+         ports:
+         - name: redis
+           containerPort: 6379
+         volumeMounts:
+         - mountPath: /data
+           name: data
+         - mountPath: /health
+           name: health
+         lifecycle:
+           {}
+       - name: sentinel
+         image: redis:6.0.7-alpine
+         imagePullPolicy: IfNotPresent
+         command:
+           - redis-sentinel
+         args:
+           - /data/conf/sentinel.conf
+         livenessProbe:
+           initialDelaySeconds: 30
+           periodSeconds: 15
+           timeoutSeconds: 15
+           successThreshold: 1
+           failureThreshold: 5
+           exec:
+             command:
+               - sh
+               - -c
+               - /health/sentinel_liveness.sh
+         resources:
+           {}
+         ports:
+           - name: sentinel
+             containerPort: 26379
+         volumeMounts:
+         - mountPath: /data
+           name: data
+         - mountPath: /health
+           name: health
+         lifecycle:
+           {}
+       volumes:
+       - name: config
+         configMap:
+           name: test-redis-ha-configmap
+       - name: health
+         configMap:
+           name: test-redis-ha-health-configmap
+           defaultMode: 0755
+   volumeClaimTemplates:
+   - metadata:
+       name: data
+       annotations:
+     spec:
+       accessModes:
+         - "ReadWriteOnce"
+       resources:
+         requests:
+           storage: "10Gi"
➜  test git:(psp_sysctl_netcoresomaxconn) ✗  (aew1-infra-core-staging:backend) helm diff upgrade --allow-unreleased --debug test .
load_plugins.go:111: [info] File (/Users/max/Library/helm/plugins/helm-2to3/completion.yaml) not provided by plugin. No plugin auto-completion possible.
load_plugins.go:111: [info] File (/Users/max/Library/helm/plugins/helm-conftest/completion.yaml) not provided by plugin. No plugin auto-completion possible.
Executing helm version
Executing helm get manifest test --namespace backend
********************

	Release was not present in Helm.  Diff will show entire contents as new.

********************
Executing helm template test . --namespace backend --validate
backend, test-redis-ha-announce-2, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-announce-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha-announce-2
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
+ spec:
+   publishNotReadyAddresses: true
+   type: ClusterIP
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
+     "statefulset.kubernetes.io/pod-name": test-redis-ha-server-2
backend, test-redis-ha, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+ spec:
+   type: ClusterIP
+   clusterIP: None
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
backend, test-redis-ha, ServiceAccount (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-serviceaccount.yaml
+ apiVersion: v1
+ kind: ServiceAccount
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     heritage: Helm
+     release: test
+     chart: redis-ha-4.12.2
+     app: test-redis-ha
backend, test-redis-ha-configmap, ConfigMap (v1) has been added:
+ # Source: test/charts/redis-ha/templates/redis-ha-configmap.yaml
+ apiVersion: v1
+ kind: ConfigMap
+ metadata:
+   name: test-redis-ha-configmap
+   namespace: backend
+   labels:
+     heritage: Helm
+     release: test
+     chart: redis-ha-4.12.2
+     app: test-redis-ha
+ data:
+   redis.conf: |
+     dir "/data"
+     port 6379
+     maxmemory 0
+     maxmemory-policy volatile-lru
+     min-replicas-max-lag 5
+     min-replicas-to-write 1
+     rdbchecksum yes
+     rdbcompression yes
+     repl-diskless-sync yes
+     save 900 1
+
+   sentinel.conf: |
+     dir "/data"
+     port 26379
+         sentinel down-after-milliseconds mymaster 10000
+         sentinel failover-timeout mymaster 180000
+         maxclients 10000
+         sentinel parallel-syncs mymaster 5
+
+   init.sh: |
+     echo "$(date) Start..."
+     HOSTNAME="$(hostname)"
+     INDEX="${HOSTNAME##*-}"
+     SENTINEL_PORT=26379
+     MASTER=''
+     MASTER_GROUP="mymaster"
+     QUORUM="2"
+     REDIS_CONF=/data/conf/redis.conf
+     REDIS_PORT=6379
+     REDIS_TLS_PORT=
+     SENTINEL_CONF=/data/conf/sentinel.conf
+     SENTINEL_TLS_PORT=
+     SERVICE=test-redis-ha
+     SENTINEL_TLS_REPLICATION_ENABLED=false
+     REDIS_TLS_REPLICATION_ENABLED=false
+     set -eu
+
+     sentinel_get_master() {
+     set +e
+         if [ "$SENTINEL_PORT" -eq 0 ]; then
+             redis-cli -h "${SERVICE}" -p "${SENTINEL_TLS_PORT}"   --tls --cacert /tls-certs/ca.crt --cert /tls-certs/redis.crt --key /tls-certs/redis.key sentinel get-master-addr-by-name "${MASTER_GROUP}" |\
+             grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}'
+         else
+             redis-cli -h "${SERVICE}" -p "${SENTINEL_PORT}"  sentinel get-master-addr-by-name "${MASTER_GROUP}" |\
+             grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}'
+         fi
+     set -e
+     }
+
+     sentinel_get_master_retry() {
+         master=''
+         retry=${1}
+         sleep=3
+         for i in $(seq 1 "${retry}"); do
+             master=$(sentinel_get_master)
+             if [ -n "${master}" ]; then
+                 break
+             fi
+             sleep $((sleep + i))
+         done
+         echo "${master}"
+     }
+
+     identify_master() {
+         echo "Identifying redis master (get-master-addr-by-name).."
+         echo "  using sentinel (test-redis-ha), sentinel group name (mymaster)"
+         echo "  $(date).."
+         MASTER="$(sentinel_get_master_retry 3)"
+         if [ -n "${MASTER}" ]; then
+             echo "  $(date) Found redis master (${MASTER})"
+         else
+             echo "  $(date) Did not find redis master (${MASTER})"
+         fi
+     }
+
+     sentinel_update() {
+         echo "Updating sentinel config.."
+         echo "  evaluating sentinel id (\${SENTINEL_ID_${INDEX}})"
+         eval MY_SENTINEL_ID="\$SENTINEL_ID_${INDEX}"
+         echo "  sentinel id (${MY_SENTINEL_ID}), sentinel grp (${MASTER_GROUP}), quorum (${QUORUM})"
+         sed -i "1s/^/sentinel myid ${MY_SENTINEL_ID}\\n/" "${SENTINEL_CONF}"
+         if [ "$SENTINEL_TLS_REPLICATION_ENABLED" = true ]; then
+             echo "  redis master (${1}:${REDIS_TLS_PORT})"
+             sed -i "2s/^/sentinel monitor ${MASTER_GROUP} ${1} ${REDIS_TLS_PORT} ${QUORUM} \\n/" "${SENTINEL_CONF}"
+         else
+             echo "  redis master (${1}:${REDIS_PORT})"
+             sed -i "2s/^/sentinel monitor ${MASTER_GROUP} ${1} ${REDIS_PORT} ${QUORUM} \\n/" "${SENTINEL_CONF}"
+         fi
+         echo "sentinel announce-ip ${ANNOUNCE_IP}" >> ${SENTINEL_CONF}
+         if [ "$SENTINEL_PORT" -eq 0 ]; then
+             echo "  announce (${ANNOUNCE_IP}:${SENTINEL_TLS_PORT})"
+             echo "sentinel announce-port ${SENTINEL_TLS_PORT}" >> ${SENTINEL_CONF}
+         else
+             echo "  announce (${ANNOUNCE_IP}:${SENTINEL_PORT})"
+             echo "sentinel announce-port ${SENTINEL_PORT}" >> ${SENTINEL_CONF}
+         fi
+     }
+
+     redis_update() {
+         echo "Updating redis config.."
+         if [ "$REDIS_TLS_REPLICATION_ENABLED" = true ]; then
+             echo "  we are slave of redis master (${1}:${REDIS_TLS_PORT})"
+             echo "slaveof ${1} ${REDIS_TLS_PORT}" >> "${REDIS_CONF}"
+             echo "slave-announce-port ${REDIS_TLS_PORT}" >> ${REDIS_CONF}
+         else
+             echo "  we are slave of redis master (${1}:${REDIS_PORT})"
+             echo "slaveof ${1} ${REDIS_PORT}" >> "${REDIS_CONF}"
+             echo "slave-announce-port ${REDIS_PORT}" >> ${REDIS_CONF}
+         fi
+         echo "slave-announce-ip ${ANNOUNCE_IP}" >> ${REDIS_CONF}
+     }
+
+     copy_config() {
+         echo "Copying default redis config.."
+         echo "  to '${REDIS_CONF}'"
+         cp /readonly-config/redis.conf "${REDIS_CONF}"
+         echo "Copying default sentinel config.."
+         echo "  to '${SENTINEL_CONF}'"
+         cp /readonly-config/sentinel.conf "${SENTINEL_CONF}"
+     }
+
+     setup_defaults() {
+         echo "Setting up defaults.."
+         echo "  using statefulset index (${INDEX})"
+         if [ "${INDEX}" = "0" ]; then
+             echo "Setting this pod as master for redis and sentinel.."
+             echo "  using announce (${ANNOUNCE_IP})"
+             redis_update "${ANNOUNCE_IP}"
+             sentinel_update "${ANNOUNCE_IP}"
+             echo "  make sure ${ANNOUNCE_IP} is not a slave (slaveof no one)"
+             sed -i "s/^.*slaveof.*//" "${REDIS_CONF}"
+         else
+             echo "Getting redis master ip.."
+             echo "  blindly assuming (${SERVICE}-announce-0) or (${SERVICE}-server-0) are master"
+             DEFAULT_MASTER="$(getent_hosts 0 | awk '{ print $1 }')"
+             echo "  identified redis (may be redis master) ip (${DEFAULT_MASTER})"
+             if [ -z "${DEFAULT_MASTER}" ]; then
+                 echo "Error: Unable to resolve redis master (getent hosts)."
+                 exit 1
+             fi
+             echo "Setting default slave config for redis and sentinel.."
+             echo "  using master ip (${DEFAULT_MASTER})"
+             redis_update "${DEFAULT_MASTER}"
+             sentinel_update "${DEFAULT_MASTER}"
+         fi
+     }
+
+     redis_ping() {
+     set +e
+         if [ "$REDIS_PORT" -eq 0 ]; then
+             redis-cli -h "${MASTER}" -p "${REDIS_TLS_PORT}"  --tls --cacert /tls-certs/ca.crt --cert /tls-certs/redis.crt --key /tls-certs/redis.key ping
+         else
+             redis-cli -h "${MASTER}" -p "${REDIS_PORT}" ping
+         fi
+     set -e
+     }
+
+     redis_ping_retry() {
+         ping=''
+         retry=${1}
+         sleep=3
+         for i in $(seq 1 "${retry}"); do
+             if [ "$(redis_ping)" = "PONG" ]; then
+                ping='PONG'
+                break
+             fi
+             sleep $((sleep + i))
+             MASTER=$(sentinel_get_master)
+         done
+         echo "${ping}"
+     }
+
+     find_master() {
+         echo "Verifying redis master.."
+         if [ "$REDIS_PORT" -eq 0 ]; then
+             echo "  ping (${MASTER}:${REDIS_TLS_PORT})"
+         else
+             echo "  ping (${MASTER}:${REDIS_PORT})"
+         fi
+         echo "  $(date).."
+         if [ "$(redis_ping_retry 3)" != "PONG" ]; then
+             echo "  $(date) Can't ping redis master (${MASTER})"
+             echo "Attempting to force failover (sentinel failover).."
+
+             if [ "$SENTINEL_PORT" -eq 0 ]; then
+                 echo "  on sentinel (${SERVICE}:${SENTINEL_TLS_PORT}), sentinel grp (${MASTER_GROUP})"
+                 echo "  $(date).."
+                 if redis-cli -h "${SERVICE}" -p "${SENTINEL_TLS_PORT}"   --tls --cacert /tls-certs/ca.crt --cert /tls-certs/redis.crt --key /tls-certs/redis.key sentinel failover "${MASTER_GROUP}" | grep -q 'NOGOODSLAVE' ; then
+                     echo "  $(date) Failover returned with 'NOGOODSLAVE'"
+                     echo "Setting defaults for this pod.."
+                     setup_defaults
+                     return 0
+                 fi
+             else
+                 echo "  on sentinel (${SERVICE}:${SENTINEL_PORT}), sentinel grp (${MASTER_GROUP})"
+                 echo "  $(date).."
+                 if redis-cli -h "${SERVICE}" -p "${SENTINEL_PORT}"  sentinel failover "${MASTER_GROUP}" | grep -q 'NOGOODSLAVE' ; then
+                     echo "  $(date) Failover returned with 'NOGOODSLAVE'"
+                     echo "Setting defaults for this pod.."
+                     setup_defaults
+                     return 0
+                 fi
+             fi
+
+             echo "Hold on for 10sec"
+             sleep 10
+             echo "We should get redis master's ip now. Asking (get-master-addr-by-name).."
+             if [ "$SENTINEL_PORT" -eq 0 ]; then
+                 echo "  sentinel (${SERVICE}:${SENTINEL_TLS_PORT}), sentinel grp (${MASTER_GROUP})"
+             else
+                 echo "  sentinel (${SERVICE}:${SENTINEL_PORT}), sentinel grp (${MASTER_GROUP})"
+             fi
+             echo "  $(date).."
+             MASTER="$(sentinel_get_master)"
+             if [ "${MASTER}" ]; then
+                 echo "  $(date) Found redis master (${MASTER})"
+                 echo "Updating redis and sentinel config.."
+                 sentinel_update "${MASTER}"
+                 redis_update "${MASTER}"
+             else
+                 echo "$(date) Error: Could not failover, exiting..."
+                 exit 1
+             fi
+         else
+             echo "  $(date) Found reachable redis master (${MASTER})"
+             echo "Updating redis and sentinel config.."
+             sentinel_update "${MASTER}"
+             redis_update "${MASTER}"
+         fi
+     }
+
+     redis_ro_update() {
+         echo "Updating read-only redis config.."
+         echo "  redis.conf set 'replica-priority 0'"
+         echo "replica-priority 0" >> ${REDIS_CONF}
+     }
+
+     getent_hosts() {
+         index=${1:-${INDEX}}
+         service="${SERVICE}-announce-${index}"
+         pod="${SERVICE}-server-${index}"
+         host=$(getent hosts "${service}")
+         if [ -z "${host}" ]; then
+             host=$(getent hosts "${pod}")
+         fi
+         echo "${host}"
+     }
+
+     mkdir -p /data/conf/
+
+     echo "Initializing config.."
+     copy_config
+
+     # where is redis master
+     identify_master
+
+     echo "Identify announce ip for this pod.."
+     echo "  using (${SERVICE}-announce-${INDEX}) or (${SERVICE}-server-${INDEX})"
+     ANNOUNCE_IP=$(getent_hosts | awk '{ print $1 }')
+     echo "  identified announce (${ANNOUNCE_IP})"
+     if [ -z "${ANNOUNCE_IP}" ]; then
+         "Error: Could not resolve the announce ip for this pod."
+         exit 1
+     elif [ "${MASTER}" ]; then
+         find_master
+     else
+         setup_defaults
+     fi
+
+     if [ "${AUTH:-}" ]; then
+         echo "Setting redis auth values.."
+         ESCAPED_AUTH=$(echo "${AUTH}" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/replace-default-auth/${ESCAPED_AUTH}/" "${REDIS_CONF}" "${SENTINEL_CONF}"
+     fi
+
+     if [ "${SENTINELAUTH:-}" ]; then
+         echo "Setting sentinel auth values"
+         ESCAPED_AUTH_SENTINEL=$(echo "$SENTINELAUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/replace-default-sentinel-auth/${ESCAPED_AUTH_SENTINEL}/" "$SENTINEL_CONF"
+     fi
+
+     echo "$(date) Ready..."
+
+   haproxy_init.sh: |
+     HAPROXY_CONF=/data/haproxy.cfg
+     cp /readonly/haproxy.cfg "$HAPROXY_CONF"
+     for loop in $(seq 1 10); do
+       getent hosts test-redis-ha-announce-0 && break
+       echo "Waiting for service test-redis-ha-announce-0 to be ready ($loop) ..." && sleep 1
+     done
+     ANNOUNCE_IP0=$(getent hosts "test-redis-ha-announce-0" | awk '{ print $1 }')
+     if [ -z "$ANNOUNCE_IP0" ]; then
+       echo "Could not resolve the announce ip for test-redis-ha-announce-0"
+       exit 1
+     fi
+     sed -i "s/REPLACE_ANNOUNCE0/$ANNOUNCE_IP0/" "$HAPROXY_CONF"
+
+     if [ "${AUTH:-}" ]; then
+         echo "Setting auth values"
+         ESCAPED_AUTH=$(echo "$AUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/REPLACE_AUTH_SECRET/${ESCAPED_AUTH}/" "$HAPROXY_CONF"
+     fi
+     for loop in $(seq 1 10); do
+       getent hosts test-redis-ha-announce-1 && break
+       echo "Waiting for service test-redis-ha-announce-1 to be ready ($loop) ..." && sleep 1
+     done
+     ANNOUNCE_IP1=$(getent hosts "test-redis-ha-announce-1" | awk '{ print $1 }')
+     if [ -z "$ANNOUNCE_IP1" ]; then
+       echo "Could not resolve the announce ip for test-redis-ha-announce-1"
+       exit 1
+     fi
+     sed -i "s/REPLACE_ANNOUNCE1/$ANNOUNCE_IP1/" "$HAPROXY_CONF"
+
+     if [ "${AUTH:-}" ]; then
+         echo "Setting auth values"
+         ESCAPED_AUTH=$(echo "$AUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/REPLACE_AUTH_SECRET/${ESCAPED_AUTH}/" "$HAPROXY_CONF"
+     fi
+     for loop in $(seq 1 10); do
+       getent hosts test-redis-ha-announce-2 && break
+       echo "Waiting for service test-redis-ha-announce-2 to be ready ($loop) ..." && sleep 1
+     done
+     ANNOUNCE_IP2=$(getent hosts "test-redis-ha-announce-2" | awk '{ print $1 }')
+     if [ -z "$ANNOUNCE_IP2" ]; then
+       echo "Could not resolve the announce ip for test-redis-ha-announce-2"
+       exit 1
+     fi
+     sed -i "s/REPLACE_ANNOUNCE2/$ANNOUNCE_IP2/" "$HAPROXY_CONF"

+     if [ "${AUTH:-}" ]; then
+         echo "Setting auth values"
+         ESCAPED_AUTH=$(echo "$AUTH" | sed -e 's/[\/&]/\\&/g');
+         sed -i "s/REPLACE_AUTH_SECRET/${ESCAPED_AUTH}/" "$HAPROXY_CONF"
+     fi
backend, test-redis-ha, Role (rbac.authorization.k8s.io) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-role.yaml
+ apiVersion: rbac.authorization.k8s.io/v1
+ kind: Role
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+ rules:
+ - apiGroups:
+     - ""
+   resources:
+     - endpoints
+   verbs:
+     - get
+ - apiGroups:
+     - 'policy'
+   resources:
+     - 'podsecuritypolicies'
+   verbs:
+     - 'use'
+   resourceNames:
+     - test-redis-ha
backend, test-redis-ha, RoleBinding (rbac.authorization.k8s.io) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-rolebinding.yaml
+ apiVersion: rbac.authorization.k8s.io/v1
+ kind: RoleBinding
+ metadata:
+   name: test-redis-ha
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+ subjects:
+ - kind: ServiceAccount
+   name: test-redis-ha
+ roleRef:
+   apiGroup: rbac.authorization.k8s.io
+   kind: Role
+   name: test-redis-ha
backend, test-redis-ha-announce-0, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-announce-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha-announce-0
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
+ spec:
+   publishNotReadyAddresses: true
+   type: ClusterIP
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
+     "statefulset.kubernetes.io/pod-name": test-redis-ha-server-0
backend, test-redis-ha-announce-1, Service (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-announce-service.yaml
+ apiVersion: v1
+ kind: Service
+ metadata:
+   name: test-redis-ha-announce-1
+   namespace: backend
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
+ spec:
+   publishNotReadyAddresses: true
+   type: ClusterIP
+   ports:
+   - name: tcp-server
+     port: 6379
+     protocol: TCP
+     targetPort: redis
+   - name: tcp-sentinel
+     port: 26379
+     protocol: TCP
+     targetPort: sentinel
+   selector:
+     release: test
+     app: redis-ha
+     "statefulset.kubernetes.io/pod-name": test-redis-ha-server-1
backend, test-redis-ha-server, StatefulSet (apps) has been added:
+ # Source: test/charts/redis-ha/templates/redis-ha-statefulset.yaml
+ apiVersion: apps/v1
+ kind: StatefulSet
+ metadata:
+   name: test-redis-ha-server
+   namespace: backend
+   labels:
+     test-redis-ha: replica
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+   annotations:
+     {}
+ spec:
+   selector:
+     matchLabels:
+       release: test
+       app: redis-ha
+   serviceName: test-redis-ha
+   replicas: 3
+   podManagementPolicy: OrderedReady
+   updateStrategy:
+     type: RollingUpdate
+   template:
+     metadata:
+       annotations:
+         checksum/init-config: ab774a0d7a7225b4e6dfe6337866183939293834dabef4ff427022fde6b1b669
+       labels:
+         release: test
+         app: redis-ha
+         test-redis-ha: replica
+     spec:
+       affinity:
+         podAntiAffinity:
+           requiredDuringSchedulingIgnoredDuringExecution:
+             - labelSelector:
+                 matchLabels:
+                   app: redis-ha
+                   release: test
+                   test-redis-ha: replica
+               topologyKey: kubernetes.io/hostname
+           preferredDuringSchedulingIgnoredDuringExecution:
+             - weight: 100
+               podAffinityTerm:
+                 labelSelector:
+                   matchLabels:
+                     app: redis-ha
+                     release: test
+                     test-redis-ha: replica
+                 topologyKey: failure-domain.beta.kubernetes.io/zone
+       securityContext:
+         fsGroup: 1000
+         runAsNonRoot: true
+         runAsUser: 1000
+         sysctls:
+         - name: net.core.somaxconn
+           value: "10000"
+       serviceAccountName: test-redis-ha
+       automountServiceAccountToken: false
+       initContainers:
+       - name: config-init
+         image: redis:6.0.7-alpine
+         imagePullPolicy: IfNotPresent
+         resources:
+           {}
+         command:
+         - sh
+         args:
+         - /readonly-config/init.sh
+         env:
+         - name: SENTINEL_ID_0
+           value: 9c038e829f6128e33b0ece23f256b3a7d3854756
+
+         - name: SENTINEL_ID_1
+           value: e8857b18f2db38508fe4cb5a24b54b26b07ec75c
+
+         - name: SENTINEL_ID_2
+           value: 2d63ff6edcd53314ed8cbc5d3bc9d6f660c977f0
+
+         volumeMounts:
+         - name: config
+           mountPath: /readonly-config
+           readOnly: true
+         - name: data
+           mountPath: /data
+

+       containers:
+       - name: redis
+         image: redis:6.0.7-alpine
+         imagePullPolicy: IfNotPresent
+         command:
+         - redis-server
+         args:
+         - /data/conf/redis.conf
+         livenessProbe:
+           initialDelaySeconds: 30
+           periodSeconds: 15
+           timeoutSeconds: 15
+           successThreshold: 1
+           failureThreshold: 5
+           exec:
+             command:
+               - sh
+               - -c
+               - /health/redis_liveness.sh
+         resources:
+           {}
+         ports:
+         - name: redis
+           containerPort: 6379
+         volumeMounts:
+         - mountPath: /data
+           name: data
+         - mountPath: /health
+           name: health
+         lifecycle:
+           {}
+       - name: sentinel
+         image: redis:6.0.7-alpine
+         imagePullPolicy: IfNotPresent
+         command:
+           - redis-sentinel
+         args:
+           - /data/conf/sentinel.conf
+         livenessProbe:
+           initialDelaySeconds: 30
+           periodSeconds: 15
+           timeoutSeconds: 15
+           successThreshold: 1
+           failureThreshold: 5
+           exec:
+             command:
+               - sh
+               - -c
+               - /health/sentinel_liveness.sh
+         resources:
+           {}
+         ports:
+           - name: sentinel
+             containerPort: 26379
+         volumeMounts:
+         - mountPath: /data
+           name: data
+         - mountPath: /health
+           name: health
+         lifecycle:
+           {}
+       volumes:
+       - name: config
+         configMap:
+           name: test-redis-ha-configmap
+       - name: health
+         configMap:
+           name: test-redis-ha-health-configmap
+           defaultMode: 0755
+   volumeClaimTemplates:
+   - metadata:
+       name: data
+       annotations:
+     spec:
+       accessModes:
+         - "ReadWriteOnce"
+       resources:
+         requests:
+           storage: "10Gi"
backend, test-redis-ha, PodSecurityPolicy (policy) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-psp.yaml
+ apiVersion: policy/v1beta1
+ kind: PodSecurityPolicy
+ metadata:
+   name: test-redis-ha
+   labels:
+     app: redis-ha
+     heritage: "Helm"
+     release: "test"
+     chart: redis-ha-4.12.2
+ spec:
+   allowPrivilegeEscalation: false
+   allowedUnsafeSysctls:
+     - net.core.somaxconn
+   fsGroup:
+     rule: 'MustRunAs'
+     ranges:
+       - min: 1000
+         max: 1000
+   hostIPC: false
+   hostNetwork: false
+   hostPID: false
+   privileged: false
+   readOnlyRootFilesystem: false
+   requiredDropCapabilities:
+     - ALL
+   runAsUser:
+     rule: 'MustRunAs'
+     ranges:
+       - min: 1000
+         max: 1000
+   seLinux:
+     rule: 'RunAsAny'
+   supplementalGroups:
+     rule: 'MustRunAs'
+     ranges:
+       - min: 1000
+         max: 1000
+   volumes:
+     - 'configMap'
+     - 'secret'
+     - 'emptyDir'
+     - 'persistentVolumeClaim'
backend, test-redis-ha-health-configmap, ConfigMap (v1) has been added:
-
+ # Source: test/charts/redis-ha/templates/redis-ha-health-configmap.yaml
+ apiVersion: v1
+ kind: ConfigMap
+ metadata:
+   name: test-redis-ha-health-configmap
+   namespace: backend
+   labels:
+     heritage: Helm
+     release: test
+     chart: redis-ha-4.12.2
+     app: test-redis-ha
+ data:
+   redis_liveness.sh: |
+     response=$(
+       redis-cli \
+         -h localhost \
+         -p 6379 \
+         ping
+     )
+     if [ "$response" != "PONG" ] && [ "${response:0:7}" != "LOADING" ] ; then
+       echo "$response"
+       exit 1
+     fi
+     echo "response=$response"
+   sentinel_liveness.sh: |
+     response=$(
+       redis-cli \
+         -h localhost \
+         -p 26379 \
+         ping
+     )
+     if [ "$response" != "PONG" ]; then
+       echo "$response"
+       exit 1
+     fi
+     echo "response=$response"
  ```
</details>